### PR TITLE
Importer permissions

### PR DESCRIPTION
**Pull request by spficklin**
Issue #474

## Type(s) of Change(s)

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] API-specific change (fix or addition to an API function)
- [ ] Updates documentation (inline or markdown files)

## Description

It was not possible to give specific users permissions to specific Importers.  There was only one permission to use them all.  Now each TripalImporter gets is own permission and a site admin can control who can use each one.

## Testing?

To test this you'll need two user accounts, one with admin permission and one without.   
1) First pull the branch and then clear the cache.   
2) Now go to the Permissions page and you should now see a set of permissions for all of the Tripal Importers.  They are grouped together with the prefix "Importer:". Set the permission for your non-admin user any importer you want to test, preferably something you have a small data set for.  Also, be sure the "Upload Files" permission is also set, otherwise user's won't be able to upload a file for importing.   Save, these settings.
3)  Log in to your site with the non-admin user and go to the page for the importer.  You will can just cut and paste the URL bar in the browser.  The page should show up.   
4)  As the non admin user try to use the loader.  It should successfully submit a job.
5)  Now go back to the site as the admin user, disable the permissions.  
6)  Return the the site as the non-admin user and the page for the importer should give an 'access denied error'.

## Additional Notes (if any):

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by laceysanderson** [comment]
This change is **not backwards compatible**... On update users who previously were able to load data via the "Load data into the Tripal site" permission now cannot. 

Instead perhaps the "Load data into the Tripal site" permission should provide permission to all loaders. If you want to restrict a user to a specific loader then give them only the specific loader permission (e.g. "Importer: Use the Chado GFF3 File Loader") and not "Load data into the Tripal site". Thoughts?

**Comment by spficklin** [comment]
It's only not backwards compatible with previous versions of Tripal v3 which is still technically "beta".  So, I'm not sure it is that critical.  But if you feel it's necessary or you feel strongly that you need a quick "switch" to enable use of all loaders at once, then I can write a function to make it work.

**Comment by laceysanderson** [comment]
I think currently it's confusing/non-intuitive. Either the Load data into the Tripal site" permission should be removed all together or act as a "quick 'switch' to enable use of all loaders at once". 

Since the second maintains backwards compatibility and saves effort for early adopters, I lean towards that.

However, an argument could be made that both specific permissions and an overriding general permission is confusing so :shrugs:. Either way works, it's just currently confusing so needs a change.

**Comment by spficklin** [comment]
Some of the non TripalImporters (e.g. Tripal Pub Importer) still use that permission.  Would you be happy if we removed that permission and set individual ones for non TripalImporer so that we have consistency?

**Comment by laceysanderson** [comment]
I'm always happy with consistency 👍

**Comment by spficklin** [comment]
Okay, I'll make the change!

**Comment by spficklin** [comment]
Okay, I have fixed this PR so that the 'load tripal data' permission is now gone.  It turns out the only data importer that did not use the new TripalImporter class was the bulk publication importer.  So, I wrote a class for it.  However, that code is a bit too complex for the TripalImporter class, so I added a new $callback, $callback_module and $callback_path static variables to the TripalImporter.  This way you can still take advantage of using the TripalImporter (i.e. consistency in permissions and menu placement).  Without this, the importer permission shows up in the 'tripal_chado' module section of the permissions page while all others show up in the 'tripal' module section. That's not very consistent.  With this, the importer now shows up with all other importers.

You'll need to clear you cache before trying this out.

**Review by laceysanderson — CHANGES_REQUESTED** [review]
Ok, I did a full functional test on the permissions and they work as expected!
I created a new test user and gave it a test role, then I did the following:
- [x] I checked that I couldn't access any of the loaders with this test user with none of the loader-specific permissions checked.
- [x] I gave the user one permission at a time and checked that they now could access each loader in succession.
- [x] I also checked that they could only access loaders they had permission to and not ones they didn't during the above test.
- [ ] Checked that all the menu items were consistent
- [x] Checked that permission names were consistent and in the same category 
- [x] Checked that the bulk pub importer still worked. Specifically I checked I could add, edit, test and submit a bulk pub importer. I also checked the job submitted ran correctly ;-)
- [x] Checked that my test user could submit a loading job

### Menu items consistent
The names of the menu items are consistent 👍 However, they are all spilled out into the main Tripal menu :-( This is due to the removal of the containing "Data Loaders" menu item:
```diff
-
-  $items['admin/tripal/loaders'] = array(
-    'title' => 'Data Loaders',
-    'description' => t('Tools facilitating data import.'),
-    'access arguments' => array('load tripal data'),
-    'type' => MENU_NORMAL_ITEM,
-    'weight' => 6
-  );
-
```
Ideally I think the access arguments would include every single individual importer permission. That way if you have permission for at least one importer then you can see a page listing the importers you have access to. However, I've never had more then one permission in the 'access arguments' so we would need to confirm it does what I think it would (eval to true if you have any one of them).

**Review by laceysanderson — APPROVED** [review]
My commit fixes the issue I found in my function + code review above.

**Comment by laceysanderson** [comment]
Now this PR just needs a code review 👍

File: tripal_chado/includes/TripalImporter/PubBulkImporter.inc [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+   * If the $use_analysis value is set above then this value indicates if the
+   * analysis should be required.
+   */
+  public static $require_analysis = FALSE;
+
+  /**
+   * Text that should appear on the button at the bottom of the importer
+   * form.
+   */
+  public static $button_text = 'Import from NCBI Taxonomy';
```
Comment (bradfordcondon) [inline tripal_chado/includes/TripalImporter/PubBulkImporter.inc]: I think this is wrong... maybe just Import?  

I'm submitting a commit with mostly spelling changes, I'll include this change with it.

**Review by bradfordcondon — APPROVED** [review]
Functionality confirmed, code looks good.

Thanks!

**Comment by spficklin** [comment]
Thanks @laceysanderson and @bradfordcondon

**Pull request merged**